### DISCUSSION TITLE
Parallel command buffers

### DISCUSSION THREAD

**Discussion by Uriopass** [bevyengine/bevy#1532] [Ideas]
I find the system local command buffers very useful, however I wish there were a concurrent alternative.

I have a lot of parallelizable queries in my application where I very rarely need to actually mutate the world, for example:

 - Add/delete a new component to my entity
 - Mutate some ressource 
 - Add/delete a component to another entity

I could put the command buffer behind a mutex, then use that mutex in the parallel query to submit commands, but I feel like this is not the most optimized way (concurrent queues could be better?) or the most ergonomic.

I also can't declare my own system local ressource (I might be wrong?) so I can't develop this myself.

Anyway, tell me what you think.

**Comment by Ratysz** [comment]
There is [`Local<T>`](https://docs.rs/bevy/0.4.0/bevy/ecs/struct.Local.html).

If you don't want to implement [`FromResources`](https://docs.rs/bevy/0.4.0/bevy/ecs/trait.FromResources.html) for the concurrent buffer, you can use [`Resources::insert_local()`](https://docs.rs/bevy/0.4.0/bevy/ecs/struct.Resources.html#method.insert_local), although it is a bit unwieldy:
```rust
let system = my_system.system();
resources.insert_local(system.id(), concurrent_buffer);
```

This will improve in Bevy 0.5.

**Reply by Uriopass** [reply]
Ah, nice! I'm not exactly sure how I'd get the same semantics as the Commands struct, which seems like it has special treatment inside the ecs. That is, run the commands at the end of each stage.

**Reply by Ratysz** [reply]
You could drain your concurrent buffers into Bevy's `Commands` at the end of the system.

**Reply by Uriopass** [reply]
That's not very practical, I would need to import both the parallel command buffer and the command buffer in all systems, and not forget to flush it. I guess it could also be a separate system that runs before the end of each stage but I'm not sure if it's possible?

**Reply by Ratysz** [reply]
It should be possible to hide that boilerplate behind a custom `SystemParam` - either implement that manually, or derive it. There is no live documentation of that, though, other than [this PR](https://github.com/bevyengine/bevy/pull/1435). Be aware that it targets main branch, not 0.4 - should still work for now, though.

> I guess it could also be a separate system that runs before the end of each stage but I'm not sure if it's possible?

It's possible, but less practical: to get another system's `Local<T>` you need to know its `SystemId`. If you store the concurrent buffers in a plain resource, you'll have to reimplement some form of the mechanism that drives `Local<T>` in order to differentiate between systems' buffers, which will likely lead to silly stuff like `Local<SystemId>`.

**Reply by Uriopass** [reply]
Using a custom system param seems like it could solve the problem. Just to be sure, is it possible to fill the commands with the parallel buffer when the system ends, so I guess it'd have to run on some destructor?  
Also, I don't mind waiting for another bevy update, I'm considering switching over from legion but I want to be sure that I can get the same ergonomics.

**Reply by TheRawMeatball** [reply]
Hmm, as a quick reminder in case you haven't seen: `Arc<parking_lot::Mutex<Commands>>` already implements SystemParam

**Reply by Ratysz** [reply]
> I guess it'd have to run on some destructor?

Most likely, yes. I haven't tried to implement something like that myself, though.

> Hmm, as a quick reminder in case you haven't seen: Arc<parking_lot::Mutex<Commands>> already implements SystemParam

This is not the same as a concurrent buffer. It can be used as a stand-in for one, but there are better ways.

**Comment by sdfgeoff** [comment]
Idle thinking:
I wonder about the feasibility of using a MPSC queue for commands is? Chuck it in a resource, get a non-mutable reference to the resource, clone the queue and submit your command. Some other system empties the queue into the Command buffer at the end of each frame

**Reply by Uriopass** [reply]
It doesn't have the same guarantees of the command buffer, but it should work, yes :-)
